### PR TITLE
Substudies added to criteria

### PR DESCRIPTION
**Pull request by alxdarksage**
They work exactly like data groups in criteria. Can be used to change schedules based on substudy membership, in particular.

See: https://sagebionetworks.jira.com/browse/BRIDGE-2324

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoCriteria.java [hunk 1]
```diff
@@ -106,6 +108,25 @@ public void setAllOfGroups(Set<String> allOfGroups) {
...
+
+    @Override
+    @DynamoDBAttribute
+    @DynamoDBTypeConverted(converter=StringSetMarshaller.class)
+    public Set<String> getAllOfSubstudyIds() {
+        return allOfSubstudyIds;
+    }
+    public void setAllOfSubstudyIds(Set<String> allOfSubstudyIds) {
+        this.allOfSubstudyIds = (allOfSubstudyIds == null) ? new HashSet<>() : allOfSubstudyIds;
+    }
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoCriteria.java:120]: Is it worth adding a test to make sure setAllOfSubstudyIds() and setNoneOfSubstudyIds() convert nulls into empty sets?

File: test/org/sagebionetworks/bridge/models/CriteriaContextTest.java [hunk 2]
```diff
@@ -42,11 +45,13 @@ public void builderWorks() {
...
+                .withUserDataGroups(TestConstants.USER_DATA_GROUPS)
+                .withUserSubstudyIds(TestConstants.USER_SUBSTUDY_IDS).build();
         
         // There are defaults
         assertEquals(CLIENT_INFO, context.getClientInfo());
         assertEquals(TestConstants.USER_DATA_GROUPS, context.getUserDataGroups());
+        assertEquals(TestConstants.USER_SUBSTUDY_IDS, context.getUserSubstudyIds());
         
         CriteriaContext copy = new CriteriaContext.Builder().withContext(context).build();
         assertEquals(CLIENT_INFO, copy.getClientInfo());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/CriteriaContextTest.java:57]: Should this part of the test also test that the copy has the substudies?

File: app/org/sagebionetworks/bridge/models/CriteriaUtils.java [hunk 3]
```diff
@@ -79,9 +86,14 @@ public static void validate(Criteria criteria, Set<String> dataGroups, Errors er
...
-        validateDataGroups(errors, dataGroups, criteria.getNoneOfGroups(), "noneOfGroups");
-        validateDataGroupNotRequiredAndProhibited(criteria, errors);
+        validateSetItemsExist(errors, "allOfGroups", dataGroups, criteria.getAllOfGroups());
+        validateSetItemsExist(errors, "noneOfGroups", dataGroups, criteria.getNoneOfGroups());
+        validateSetItemsDoNotOverlap(errors, "allOfGroups", criteria.getAllOfGroups(), criteria.getNoneOfGroups());
+        
+        validateSetItemsExist(errors, "allOfSubstudyIds", substudyIds, criteria.getAllOfSubstudyIds());
+        validateSetItemsExist(errors, "noneOfSubstudyIds", substudyIds, criteria.getNoneOfSubstudyIds());
+        validateSetItemsDoNotOverlap(errors, "allOfSubstudyIds", criteria.getAllOfSubstudyIds(),
+                criteria.getNoneOfSubstudyIds());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/CriteriaUtils.java:97]: Should we add the substudy equivalents of tests validateDataGroupCannotBeWrong() and validateDataGroupNotBothRequiredAndProhibited()?

File: app/org/sagebionetworks/bridge/models/RequestInfo.java [hunk 4]
```diff
@@ -146,6 +157,7 @@ public Builder copyOf(RequestInfo requestInfo) {
                 withUserAgent(requestInfo.getUserAgent());
                 withLanguages(requestInfo.getLanguages());
                 withUserDataGroups(requestInfo.getUserDataGroups());
+                withUserSubstudyIds(requestInfo.getUserSubstudyIds());
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/models/RequestInfo.java:160]: Should we add a unit test for copyOf()?

File: test/org/sagebionetworks/bridge/models/schedules/CriteriaScheduleStrategyTest.java [hunk 5]
```diff
@@ -286,9 +293,9 @@ public void validatesInvalidScheduleCriteria() {
             Validate.entityThrowingException(VALIDATOR, PLAN);     
             fail("Should have thrown exception");
         } catch(InvalidEntityException e) {
-            assertError(e, "strategy.scheduleCriteria[1].criteria.allOfGroups", 0, " 'group1' is not in enumeration: <no data groups declared>");
-            assertError(e, "strategy.scheduleCriteria[2].criteria.noneOfGroups", 0, " 'group2' is not in enumeration: <no data groups declared>");
-            assertError(e, "strategy.scheduleCriteria[2].criteria.noneOfGroups", 1, " 'group1' is not in enumeration: <no data groups declared>");
+            assertError(e, "strategy.scheduleCriteria[1].criteria.allOfGroups", 0, " 'group1' is not in enumeration: <empty>");
+            assertError(e, "strategy.scheduleCriteria[2].criteria.noneOfGroups", 0, " 'group2' is not in enumeration: <empty>");
+            assertError(e, "strategy.scheduleCriteria[2].criteria.noneOfGroups", 1, " 'group1' is not in enumeration: <empty>");
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/models/schedules/CriteriaScheduleStrategyTest.java:300]: Should we add substudies to this?

File: test/org/sagebionetworks/bridge/play/controllers/UserProfileControllerTest.java [hunk 6]
```diff
@@ -302,11 +303,12 @@ public void validDataGroupsCanBeAdded() throws Exception {
         assertEquals(ID, participant.getId());
         assertEquals(dataGroupSet, participant.getDataGroups());
         assertEquals("First", participant.getFirstName());
-        
+        assertEquals(TestConstants.USER_SUBSTUDY_IDS, participant.getSubstudyIds());
         assertEquals(dataGroupSet, contextCaptor.getValue().getUserDataGroups());
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/play/controllers/UserProfileControllerTest.java:307]: Should we also test that the captured context contains the substudies?

File: app/org/sagebionetworks/bridge/services/ParticipantService.java [hunk 7]
```diff
@@ -295,7 +299,8 @@ public StudyParticipant getParticipant(Study study, Account account, boolean inc
                         .withHealthCode(account.getHealthCode())
                         .withClientInfo(requestInfo.getClientInfo())
                         .withLanguages(requestInfo.getLanguages())
-                        .withUserDataGroups(requestInfo.getUserDataGroups())
+                        .withUserDataGroups(account.getDataGroups())
+                        .withUserSubstudyIds(assoc.getSubstudyIdsVisibleToCaller())
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ParticipantService.java:303]: Should the getStudyParticipant() unit test verify the CriteriaContext includes substudies?

File: app/org/sagebionetworks/bridge/services/ParticipantService.java [hunk 8]
```diff
@@ -859,7 +864,7 @@ private void updateRequestContext(ExternalIdentifier externalId) {
     }
      
     
-    private CriteriaContext getCriteriaContextForParticipant(Study study, StudyParticipant participant) {
+    protected CriteriaContext getCriteriaContextForParticipant(Study study, StudyParticipant participant) {
```
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/services/ParticipantService.java]: Who uses this, and should the tests be verifying that it contains the substudies?

File: test/org/sagebionetworks/bridge/services/SubpopulationServiceTest.java [hunk 9]
```diff
@@ -157,6 +161,7 @@ public void createSubpopulation() {
         verify(subpopDao).createSubpopulation(subpop);
         verify(studyConsentService).addConsent(eq(result.getGuid()), any());
         verify(studyConsentService).publishConsent(study, result, CONSENT_CREATED_ON);
+        verify(substudyService).getSubstudyIds(TEST_STUDY);
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/SubpopulationServiceTest.java:164]: Should this be added to updateSubpopulation()?

File: test/org/sagebionetworks/bridge/services/SubstudyServiceTest.java [hunk 10]
```diff
@@ -60,6 +61,16 @@ public void getSubstudy() {
         verify(substudyDao).getSubstudy(TestConstants.TEST_STUDY, "id");
     }
     
+    @Test
+    public void getSubstudyIds() {
+        SUBSTUDIES.get(0).setId("substudyA");
+        SUBSTUDIES.get(1).setId("substudyB");
```
Comment (DwayneJengSage) [inline test/org/sagebionetworks/bridge/services/SubstudyServiceTest.java]: We shouldn't be modifying constants. If these are truly constants, they should be statically initialized and then never changed. If these aren't constants, we should create them in before() to ensure they're being reset with each test.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: app/org/sagebionetworks/bridge/dynamodb/DynamoCriteria.java (hunk 1 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/dynamodb/DynamoCriteria.java:120]: Added test for the four set fields.

File: test/org/sagebionetworks/bridge/models/CriteriaContextTest.java (hunk 2 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/CriteriaContextTest.java:57]: Yes. Changed to test for all five of the fields that are set in the context object.

File: app/org/sagebionetworks/bridge/models/CriteriaUtils.java (hunk 3 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/CriteriaUtils.java:97]: Added (found bug!)

File: app/org/sagebionetworks/bridge/models/RequestInfo.java (hunk 4 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/models/RequestInfo.java:160]: Added

File: test/org/sagebionetworks/bridge/models/schedules/CriteriaScheduleStrategyTest.java (hunk 5 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/models/schedules/CriteriaScheduleStrategyTest.java:300]: Added

File: test/org/sagebionetworks/bridge/play/controllers/UserProfileControllerTest.java (hunk 6 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/play/controllers/UserProfileControllerTest.java:307]: Added

File: app/org/sagebionetworks/bridge/services/ParticipantService.java (hunk 7 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ParticipantService.java:303]: Added

File: app/org/sagebionetworks/bridge/services/ParticipantService.java (hunk 8 above)
Comment (alxdarksage) [reply app/org/sagebionetworks/bridge/services/ParticipantService.java]: Seems like this should have been reverted after thinking about accessing it from a test. Returned it to private visibility.

File: test/org/sagebionetworks/bridge/services/SubpopulationServiceTest.java (hunk 9 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/SubpopulationServiceTest.java:164]: Yes, added

File: test/org/sagebionetworks/bridge/services/SubstudyServiceTest.java (hunk 10 above)
Comment (alxdarksage) [reply test/org/sagebionetworks/bridge/services/SubstudyServiceTest.java]: Created new list of subsidies for this test.

**Review by DwayneJengSage — APPROVED** [review]
